backend/src/search/SearchIndexBuilder.ts

```diff
@@ -0,0 +1,342 @@
+        indexValue: stringifiedValue,
+      });
+    }
+  }
+}
+
+/**
+ * Translates an entity-type to an index-entry and records it in an entries collection.
+ * @param entityType The entity-type to translates.
+ * @param type The type of entity the translation affects. This is either a node-type or an edge-type.
+ * @param entries The collection of index entries.
+ * @returns The translated entity-type.
+ */
+function recordEntityType(
+  entityType: EntityType,
+  type: 'node-type' | 'edge-type',
+  entries: IndexEntry[]
+): void {
+  // We store an entity type as index entry in the following way:
+  // The id of the entry is the name of the type,
+  // There is a single entry-property thats name is `property`and thats value is the concatenated names of
+  // all properties the entity-type contains.
+  // For example:
+  // A node of type Person consists of the properties 'name', 'born', 'address'.
+  // The index entry looks like:
+  // { id: 'Person', indexValue: '[name, born, address]' }
+
+  // For an entity-type, we want to search in the name of the entity-type
+  // and in the name of the properties, so we have to insert two entries into the index.
+  const entityTypeNameEntry = {
+    entityType: type,
+    id: entityType.name,
+    indexKey: 'id',
+    indexValue: entityType.name,
+  };
+
+  entries.push(entityTypeNameEntry);
+
+  if (entityType.properties.length > 0) {
+    const entityTypePropertiesEntry = {
+      entityType: type,
+      id: entityType.name,
+      indexKey: 'properties.properties',
+      indexValue: flattenArray(entityType.properties.map((prop) => prop.name)),
+    };
+
+    entries.push(entityTypePropertiesEntry);
+  }
+}
+
+/**
+ * Translates an edge-type to an index-entry and records it in an entries collection.
+ * @param edgeType The edge-type to translate.
+ * @param entries The collection of index entries.
+ * @returns The translated edge-type.
+ */
+function recordEdgeType(edgeType: EdgeType, entries: IndexEntry[]): void {
+  const nodeEntityEntry: IndexEntry = {
+    entityType: 'edge-type',
+    id: edgeType.name,
+    indexKey: 'entity-type',
+    indexValue: 'edge',
+  };
+
+  entries.push(nodeEntityEntry);
+
+  recordEntityType(edgeType, 'edge-type', entries);
+}
+
+/**
+ * Translates a node-type to an index-entry and records it in an entries collection.
+ * @param nodeType The node-type to translate.
+ * @param entries The collection of index entries.
+ * @returns The translated node-type.
+ */
+function recordNodeType(nodeType: NodeType, entries: IndexEntry[]): void {
+  const nodeEntityEntry: IndexEntry = {
+    entityType: 'node-type',
+    id: nodeType.name,
+    indexKey: 'entity-type',
+    indexValue: 'node',
+  };
+
+  entries.push(nodeEntityEntry);
+
+  recordEntityType(nodeType, 'node-type', entries);
+}
+
+/**
+ * A search index builder that can be used to build a search index from a complete dataset.
+ */
+@Injectable()
+export class SearchIndexBuilder {
```

Comment:
Move the functions above in another file, they take too much space.

Reply:
I refactored the  recording stuff to a dedicated type and moved the rather generic functions to own files in the utils dir.